From 891d2d42484880d41b25551de3c315ba70a63225 Mon Sep 17 00:00:00 2001
From: Ian Dannapel <ian.dannapel@iris-sensing.com>
Date: Fri, 13 May 2022 15:27:16 +0200
Subject: [PATCH] imx8mp-evk.dts: Add epc660 and tc358746 nodes to dts

Remove dtb from Makefile which depends on the previous ov5640 camera
node being present in the -evk devicetree

Signed-off-by: Ian Dannapel <ian.dannapel@iris-sensing.com>
---
 arch/arm64/boot/dts/freescale/Makefile       |  2 +-
 arch/arm64/boot/dts/freescale/imx8mp-evk.dts | 81 +++++++++++++++-----
 2 files changed, 61 insertions(+), 22 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 9a7319c6b4e3..21724f244dd5 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -109,7 +109,7 @@ dtb-$(CONFIG_ARCH_MXC) += imx8mp-evk.dtb imx8mp-evk-rm67191.dtb imx8mp-evk-it626
 			  imx8mp-evk-basler-ov2775.dtb imx8mp-evk-dual-basler.dtb \
 			  imx8mp-evk-dual-ov2775.dtb imx8mp-evk-spdif-lb.dtb \
 			  imx8mp-evk-sof-wm8960.dtb imx8mp-evk-dsp.dtb \
-			  imx8mp-evk-os08a20-ov5640.dtb imx8mp-evk-os08a20.dtb \
+			  imx8mp-evk-os08a20.dtb \
 			  imx8mp-evk-dual-os08a20.dtb \
 			  imx8mp-evk-iqaudio-dacplus.dtb imx8mp-evk-iqaudio-dacpro.dtb imx8mp-evk-hifiberry-dacplus.dtb \
 			  imx8mp-evk-hifiberry-dac2.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx8mp-evk.dts b/arch/arm64/boot/dts/freescale/imx8mp-evk.dts
index c451d4b2872b..83da01155a5e 100644
--- a/arch/arm64/boot/dts/freescale/imx8mp-evk.dts
+++ b/arch/arm64/boot/dts/freescale/imx8mp-evk.dts
@@ -190,6 +190,13 @@ usb3_data_ss: endpoint {
 			};
 		};
 	};
+
+	clk_serialize: clock_serialize {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <32000000>;
+		clock-output-names = "32m_serialize";
+	};
 };
 
 &flexspi {
@@ -518,29 +525,59 @@ it6263_in: endpoint {
 		};
 	};
 
-	ov5640_0: ov5640_mipi@3c {
-		compatible = "ovti,ov5640";
-		reg = <0x3c>;
+	epc660_mipi: epc660_mipi@22 {
+		compatible = "epc660";
+		reg = <0x22>;
+		status = "okay";
 		pinctrl-names = "default";
-		pinctrl-0 = <&pinctrl_csi0_pwn>, <&pinctrl_csi0_rst>, <&pinctrl_csi_mclk>;
-		clocks = <&clk IMX8MP_CLK_IPP_DO_CLKO2>;
-		clock-names = "xclk";
-		assigned-clocks = <&clk IMX8MP_CLK_IPP_DO_CLKO2>;
-		assigned-clock-parents = <&clk IMX8MP_CLK_24M>;
-		assigned-clock-rates = <24000000>;
-		csi_id = <0>;
-		powerdown-gpios = <&gpio2 11 GPIO_ACTIVE_HIGH>;
+		pinctrl-0 = <&pinctrl_csi0_rst>, <&pinctrl_csi_mclk>;
 		reset-gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
-		mclk = <24000000>;
-		mclk_source = <0>;
-		mipi_csi;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			epc660_parallel_out: endpoint {
+				bus-type = <5>;
+				bus-width = <12>;
+				hsync-active = <0>;
+				vsync-active = <0>;
+				field-even-active = <0>;
+				pclk-sample = <1>;
+				remote-endpoint = <&tc358748_parallel_in>;
+			};
+		};
+	};
+
+	tc358748_adapter: csi-bridge@e {
+		compatible = "toshiba,tc358746";
+		reg = <0x0e>;
 		status = "okay";
 
-		port {
-			ov5640_mipi_0_ep: endpoint {
-				remote-endpoint = <&mipi_csi0_ep>;
+		clocks = <&clk_serialize>;
+		default-input = <0>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		port@0 {
+			reg = <0>;
+			tc358748_parallel_in: endpoint {
+				bus-type = <5>;
+				bus-width = <12>;
+				hsync-active = <0>;
+				vsync-active = <0>;
+				field-even-active = <0>;
+				remote-endpoint = <&epc660_parallel_out>;
+			};
+		};
+		port@1 {
+			reg = <1>;
+			tc358748_mipi_out: endpoint {
 				data-lanes = <1 2>;
-				clock-lanes = <0>;
+				link-frequencies = /bits/ 64 <216000000>;
+				remote-endpoint = <&mipi_csi0_ep>;
 			};
 		};
 	};
@@ -1330,11 +1367,13 @@ &mipi_csi_0 {
 	port@0 {
 		reg = <0>;
 		mipi_csi0_ep: endpoint {
-			remote-endpoint = <&ov5640_mipi_0_ep>;
 			data-lanes = <2>;
-			csis-hs-settle = <13>;
-			csis-clk-settle = <2>;
+			clock-lanes = <1>;
+			remote-endpoint = <&tc358748_mipi_out>;
+			csis-hs-settle = <3>;
+			csis-clk-settle = <0>;
 			csis-wclk;
+			link-frequencies = /bits/ 64 <216000000>;
 		};
 	};
 };
-- 
2.39.1

